### PR TITLE
plugins: nordic-hid: Add peers cache support

### PR DESCRIPTION
**Pull request by MarekPieta**
PR introduces configuration channel peers cache functionality support. The peers cache allows host tools to get informed about the state of nRF Desktop peripherals connected through dongle without need of complete rediscovery.

The PR also introduces a few small changes needed to integrate the peers cache functionality.

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 1]
```diff
@@ -419,14 +423,75 @@ fu_nordic_hid_cfg_channel_cmd_receive(FuNordicHidCfgChannel *self,
 }
 
 static gboolean
-fu_nordic_hid_cfg_channel_add_peers(FuNordicHidCfgChannel *self, GError **error)
+is_cached_peer_connected(guint8 peer_cache_val)
```
Comment (hughsie) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:426]: can this have a fu_nordic_hid_cfg_ prefix please.

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 2]
```diff
@@ -419,14 +423,75 @@ fu_nordic_hid_cfg_channel_cmd_receive(FuNordicHidCfgChannel *self,
...
+{
+	return (peer_cache_val % 2) != 0;
+}
+
+static void
+fu_nordic_hid_cfg_channel_add_peer(FuNordicHidCfgChannel *self, guint8 peer_id)
+{
+	g_autoptr(GError) error_peer = NULL;
```
Comment (hughsie) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:434]: error_local would be more typical

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 3]
```diff
@@ -419,14 +423,75 @@ fu_nordic_hid_cfg_channel_cmd_receive(FuNordicHidCfgChannel *self,
...
+	g_autoptr(FuNordicHidCfgChannel) peer = NULL;
+
+	peer = fu_nordic_hid_cfg_channel_new(peer_id, self);
+
+	/* ensure that the general quirk content for Nordic HID devices is applied */
+	fu_device_add_instance_id(FU_DEVICE(peer), "HIDRAW\\VEN_1915");
+
+	if (!fu_device_setup(FU_DEVICE(peer), &error_peer)) {
+		g_debug("failed to discover peer 0x%02x: %s", peer_id, error_peer->message);
+	} else {
```
Comment (hughsie) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:444]: maybe return from the error and remove the else clause?

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 4]
```diff
@@ -419,14 +423,75 @@ fu_nordic_hid_cfg_channel_cmd_receive(FuNordicHidCfgChannel *self,
...
+						    guint8 peers_cache[PEERS_CACHE_LEN])
+{
+	for (guint i = 0; i < PEERS_CACHE_LEN; i++) {
+		guint8 peer_id = i + 1;
+
+		if (peers_cache == NULL || !is_cached_peer_connected(peers_cache[i])) {
+			fu_nordic_hid_cfg_channel_remove_peer(self, peer_id);
+			if (peers_cache != NULL) {
+				self->peers_cache[i] = peers_cache[i];
+			}
```
Comment (hughsie) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:481]: no need for {}

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 5]
```diff
@@ -436,61 +501,226 @@ fu_nordic_hid_cfg_channel_add_peers(FuNordicHidCfgChannel *self, GError **error)
...
 	if (!fu_nordic_hid_cfg_channel_cmd_receive(self, CONFIG_STATUS_SUCCESS, res, error))
 		return FALSE;
 
+	*cmd_supported = TRUE;
+	return TRUE;
+}
+
+static gboolean
+fu_nordic_hid_cfg_channel_get_next_peer_id_cmd(FuNordicHidCfgChannel *self,
+					       GError **error,
```
Comment (hughsie) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:523]: GError is normally the last parameter

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 6]
```diff
@@ -436,61 +501,226 @@ fu_nordic_hid_cfg_channel_add_peers(FuNordicHidCfgChannel *self, GError **error)
...
+		return FALSE;
+
+	*peer_id = res->data[8];
+
+	return TRUE;
+}
+
+static gboolean
+fu_nordic_hid_cfg_channel_read_peers_cache_cmd(FuNordicHidCfgChannel *self,
+					       GError **error,
```
Comment (hughsie) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:547]: same, last parameter pls

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 7]
```diff
@@ -436,61 +501,226 @@ fu_nordic_hid_cfg_channel_add_peers(FuNordicHidCfgChannel *self, GError **error)
...
+	*cmd_supported = FALSE;
+
+	if (!fu_nordic_hid_cfg_channel_cmd_send(self,
+						NULL,
+						NULL,
+						CONFIG_STATUS_GET_PEERS_CACHE,
+						NULL,
+						0,
+						error))
+		return FALSE;
```
Comment (hughsie) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:563]: maybe add a g_prefix_error here for our future selves

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 8]
```diff
@@ -436,61 +501,226 @@ fu_nordic_hid_cfg_channel_add_peers(FuNordicHidCfgChannel *self, GError **error)
...
+
+	if (fu_nordic_hid_cfg_channel_cmd_receive(self,
+						  CONFIG_STATUS_DISCONNECTED,
+						  res,
+						  &error_local)) {
+		/* Configuration channel peers cache not supported. */
+		return TRUE;
+	}
+
+	/* Configuration channel peer caching available */
```
Comment (hughsie) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:573]: lowercase c pls

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 9]
```diff
@@ -436,61 +501,226 @@ fu_nordic_hid_cfg_channel_add_peers(FuNordicHidCfgChannel *self, GError **error)
...
+						NULL,
+						0,
+						error))
+		return FALSE;
+
+	if (fu_nordic_hid_cfg_channel_cmd_receive(self,
+						  CONFIG_STATUS_DISCONNECTED,
+						  res,
+						  &error_local)) {
+		/* Configuration channel peers cache not supported. */
```
Comment (hughsie) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:569]: configuration

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 10]
```diff
@@ -436,61 +501,226 @@ fu_nordic_hid_cfg_channel_add_peers(FuNordicHidCfgChannel *self, GError **error)
...
+
+	return TRUE;
+}
+
+static gboolean
+fu_nordic_hid_cfg_channel_update_peers(FuNordicHidCfgChannel *self,
+				       GError **error,
+				       guint8 peers_cache[PEERS_CACHE_LEN])
+{
+	g_autoptr(GError) error_local = NULL;
```
Comment (hughsie) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:597]: g_autoptr() usually come last in the variable block

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 11]
```diff
@@ -436,61 +501,226 @@ fu_nordic_hid_cfg_channel_add_peers(FuNordicHidCfgChannel *self, GError **error)
...
+		return FALSE;
+
+	*cmd_supported = TRUE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_nordic_hid_cfg_channel_update_peers(FuNordicHidCfgChannel *self,
+				       GError **error,
```
Comment (hughsie) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:594]: last param pls

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 12]
```diff
@@ -436,61 +501,226 @@ fu_nordic_hid_cfg_channel_add_peers(FuNordicHidCfgChannel *self, GError **error)
...
+static gboolean
+fu_nordic_hid_cfg_channel_setup_peers(FuNordicHidCfgChannel *self, GError **error)
+{
+	gboolean peers_supported = FALSE;
+	gboolean peers_cache_supported = FALSE;
+	guint8 peers_cache[PEERS_CACHE_LEN] = {0x00};
+
+	if (self->peer_id != 0)
+		/* device connected through dongle cannot support peers. */
+		return TRUE;
```
Comment (hughsie) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:658]: needs braces

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c (hunk 9 above)
Comment (mkapala-nordic) [reply plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:569]: i think that there should not be also period at the end, same in lines: 609, 624, 657, 682, 716

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c (hunk 3 above)
Comment (MarekPieta) [reply plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:444]: sure

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c (hunk 10 above)
Comment (MarekPieta) [reply plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:597]: Aligned

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c (hunk 12 above)
Comment (MarekPieta) [reply plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:658]: Thanks, I missed this while adding a comment

**Comment by MarekPieta** [comment]
Hi @hughsie,
Added the requested changes. Please rereview.

File: plugins/nordic-hid/fu-nordic-hid-cfg-channel.c [hunk 13]
```diff
@@ -419,78 +423,317 @@ fu_nordic_hid_cfg_channel_cmd_receive(FuNordicHidCfgChannel *self,
...
+			if (peers_cache != NULL)
+				self->peers_cache[i] = peers_cache[i];
+		}
+	}
+}
+
+static gboolean
+fu_nordic_hid_cfg_channel_index_peers_cmd(FuNordicHidCfgChannel *self,
+					  GError **error,
+					  gboolean *cmd_supported)
```
Comment (mkapala-nordic) [inline plugins/nordic-hid/fu-nordic-hid-cfg-channel.c:489]: I think you aligned in every other function that `error` is last, consider aligning it here also

**Review by mkapala-nordic — APPROVED** [review]

**Comment by hughsie** [comment]
@MarekPieta I wondered if you could put some of this knowledge into the plugin README as well?

**Comment by MarekPieta** [comment]
Hi @hughsie,
Could we update the readme in a separate PR? I would add a new section related to handling configuration channel device discovery - I think it's needed to start describing peers caching. I would also add the links to documentation in nRF Connect SDK (I would like to slightly update it too).

**Comment by hughsie** [comment]
@MarekPieta sure, that's fine -- but I think we really need to do that soon -- the code is getting quite complicated. Do you want me to merge this PR before 1.9.5 (release is today!) and then you work on a different PR to merge after that?

**Comment by MarekPieta** [comment]
@hughsie, it would be great if we could merge it today, so it's included in the release.
The documentation related to configuration channel protocol handling in the plugin is not added at this point, so we could treat it as a general documentation improvement. I think I may not be able to finalize the doc alignment today, so probably it will be added after the release (I will try to prioritize this one to add it soon).

**Pull request merged**